### PR TITLE
[Data] Add `Datasource.on_write_start`

### LINKED ISSUE #36690: [Data] `SLOW_DOWN` AWS error when writing to S3
### What happened + What you expected to happen

I tried to write my Dataset to S3, but I got an AWS error. This error occurred when I increased the number of blocks (128, 256, 512).

```
INFO:common.aws:Finished writing dataset to S3
INFO:interaction_data.transformer:CP interaction processing pipeline successfully ran.
(Write pid=877, ip=10.103.21.191) 2023-06-20 14:01:51,280	INFO worker.py:861 -- Task failed with retryable exception: TaskID(578efd91c48bdf36ffffffffffffffffffffffff03000000). [repeated 22x across cluster] (Ray deduplicates logs by default. Set RAY_DEDUP_LOGS=0 to disable log deduplication, or see https://docs.ray.io/en/master/ray-observability/ray-logging.html#log-deduplication for more options.)
(Write pid=877, ip=10.103.21.191) Traceback (most recent call last): [repeated 22x across cluster]
(Write pid=877, ip=10.103.21.191)   File "python/ray/_raylet.pyx", line 1207, in ray._raylet.execute_dynamic_generator_and_store_task_outputs [repeated 22x across cluster]
(Write pid=877, ip=10.103.21.191)   File "python/ray/_raylet.pyx", line 3427, in ray._raylet.CoreWorker.store_task_outputs [repeated 22x across cluster]
(Write pid=877, ip=10.103.21.191)   File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/data/_internal/execution/operators/map_operator.py", line 389, in _map_task [repeated 22x across cluster]
(Write pid=877, ip=10.103.21.191)     for b_out in fn(iter(blocks), ctx): [repeated 22x across cluster]
(Write pid=877, ip=10.103.21.191)   File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/data/_internal/execution/legacy_compat.py", line 305, in do_map [repeated 22x across cluster]
(Write pid=877, ip=10.103.21.191)     yield from block_fn(blocks, ctx, *fn_args, **fn_kwargs) [repeated 22x across cluster]
(Write pid=877, ip=10.103.21.191)   File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/data/dataset.py", line 2762, in write_fn_wrapper [repeated 22x across cluster]
(Write pid=877, ip=10.103.21.191)     return write_fn(blocks, ctx) [repeated 22x across cluster]
(Write pid=877, ip=10.103.21.191)   File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/data/_internal/planner/write.py", line 20, in fn [repeated 22x across cluster]
(Write pid=877, ip=10.103.21.191)     {"write_result": [datasource.write(blocks, ctx, **write_args)]} [repeated 22x across cluster]
(Write pid=877, ip=10.103.21.191)   File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/data/datasource/file_based_datasource.py", line 295, in write [repeated 22x across cluster]
(Write pid=877, ip=10.103.21.191)     filesystem.create_dir(tmp, recursive=True) [repeated 22x across cluster]
(Write pid=877, ip=10.103.21.191)   File "pyarrow/_fs.pyx", line 593, in pyarrow._fs.FileSystem.create_dir [repeated 22x across cluster]
(Write pid=877, ip=10.103.21.191)   File "pyarrow/error.pxi", line 115, in pyarrow.lib.check_status [repeated 22x across cluster]
(Write pid=877, ip=10.103.21.191) OSError: When creating key 'item-to-item-recs/202306201335_session_interactions/' in bucket 'reverb-production-anyscale': AWS Error SLOW_DOWN during PutObject operation: Please reduce your request rate. [repeated 8x across cluster]
(Write pid=1131, ip=10.103.25.17) OSError: When creating key 'item-to-item-recs/' in bucket 'reverb-production-anyscale': AWS Error SLOW_DOWN during PutObject operation: Please reduce your request rate. [repeated 13x across cluster]
```

### Versions / Dependencies

Master

### Reproduction script

```
def write_ray_dataset_to_s3(path: str, dataset: ray.data.Dataset) -> None:
    """
    Saves a provided Ray dataset into a parquet file to a provided s3 path.

    Args:
        path (str): S3 path where the parquet file will be written.
        dataset (ray.data.Dataset): Ray dataset to be written to the parquet file.

    Returns:
        None.
    """
    session = boto3.Session(profile_name=get_storage_options().get("profile"))

    fs = pyarrow.fs.S3FileSystem(
        region=session.region_name,
        access_key=session.get_credentials().access_key,
        secret_key=session.get_credentials().secret_key,
        session_token=session.get_credentials().token,
    )

    dataset.write_parquet(path, filesystem=fs)

    logger.info("Finished writing dataset to S3")
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

Currently, we attempt to create a directory in every write task. This can cause rate limiting issues with S3. To address this problem, this PR adds a `on_write_start` method that is executed once per write job.

## Related issue number

Fixes #36690

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by ericl** [comment]
Nice find!

**Comment by bveeramani** [comment]
@c21 Here's the currently failing test:

```
  | # Ensure that directory was created for only the non-empty dataset.
  | if fs is None:
  | assert not os.path.isdir(all_empty_path)
  | assert os.path.isdir(some_empty_path)
  | # Only files for the non-empty blocks should be created.
  | file_list = os.listdir(some_empty_path)
  | file_list.sort()
  | assert file_list == [
  | f"{some_empty_key}_00000{i}_000000.parquet" for i in range(2)
  | ]
  | else:
  | >           assert (
  | fs.get_file_info(_unwrap_protocol(all_empty_path)).type
  | == pa.fs.FileType.NotFound
  | )
  | E           assert <FileType.Directory: 3> == <FileType.NotFound: 0>
  | E             +<FileType.Directory: 3>
  | E             -<FileType.NotFound: 0>
  |  
  | python/ray/data/tests/test_parquet.py:861: AssertionError
```

**Review by c21 — COMMENTED** [review]
The change with `on_write_start` LGTM. As discussed offline, let's delete the staging directory if no file is written.

File: python/ray/data/datasource/file_based_datasource.py [hunk 1]
```diff
@@ -367,6 +379,20 @@ def write(
         # succeeds.
         return "ok"
 
+    def on_write_complete(
```
Comment (c21) [inline python/ray/data/datasource/file_based_datasource.py:382]: let's only delete the directory created by ourselves, but not from users. Should we guard it with `try_create_dir = True`?

**Review by c21 — APPROVED** [review]

**Pull request merged**